common/json.c

```diff
@@ -301,6 +301,90 @@ const jsmntok_t *json_get_arr(const jsmntok_t tok[], size_t index)
 	return NULL;
 }
 
+/* Finds the end of a JSMN token.  */
+static const jsmntok_t *
+find_jsmn_bounds(const jsmntok_t *toks, const jsmntok_t *toks_end, int size)
+{
+	const jsmntok_t *p = toks;
+	const jsmntok_t *e = toks_end;
+	int i;
+
+	for (i = 0; i < size && p < e; ++i) {
+		if (p->size == 0)
+			++p;
+		else
+			p = find_jsmn_bounds(p + 1, e, p->size);
+	}
+
+	return p;
+}
+
+/* Validate that the JSMN result is not weird, e.g.
+ { "" "" "" }, which JSMN will accept incorectly.
+*/
+static bool
+validate_jsmn_result(const jsmntok_t *toks, const jsmntok_t *toks_end)
+{
+	const jsmntok_t *p = toks;
+	const jsmntok_t *e = toks_end;
+
+	while (p < e) {
+		if (p->type == JSMN_STRING
+		  || p->type == JSMN_PRIMITIVE
+		  || p->type == JSMN_UNDEFINED) {
+			/* Top-level, non-object-owned should have 0 size.  */
+			if (p->size != 0)
+				return false;
+			++p;
+		} else if (p->type == JSMN_ARRAY) {
+			const jsmntok_t *arrayend;
+			arrayend = find_jsmn_bounds(p + 1, e, p->size);
+			/* Invalid array.  */
+			if (p->size != 0 && arrayend == p + 1)
+				return false;
+			if (!validate_jsmn_result(p + 1, arrayend))
+				return false;
+			p = arrayend;
+		} else if (p->type == JSMN_OBJECT) {
+			const jsmntok_t *pp;
+			const jsmntok_t *objend;
+			int count = 0;
+			objend = find_jsmn_bounds(p + 1, e, p->size);
+
+			/* Invalid object.  */
+			if (p->size != 0 && objend == p + 1)
+				return false;
+
+			pp = p + 1;
+			while (pp < objend) {
+				/* Check the key.  */
+				if (pp->type != JSMN_STRING)
+					return false;
+				if (pp->size != 1)
+					return false;
```

Comment:
Don't quite follow why this needs to be `1`, while `JSMN_STRING` instances in the top-level must have size `0` (see line 336).

Reply:
libjsmn parses the text `"foo": 2` as:

```
[0]
        type = JSMN_STRING
        start = 1
        end = 4
        size = 1
[1]
        type = JSMN_PRIMITIVE
        start = 7
        end = 8
        size = 0
```

In fact, an "object" in libjsmn is not a dictionary.  It is a *sequence* of *key-value pairs* that is delimited by `{` and `}`.  A "key-value pair" is a `JSMN_STRING` key that contains (`size == 1`) a value.

Indeed, you could write `{ "", "" }` in libjsmn and it would accept that, creating:

```
[0]
        type = JSMN_OBJECT
        start = 0
        end = 10
        size = 2
[1]
        type = JSMN_STRING
        start = 3
        end = 3
        size = 0
[2]
        type = JSMN_STRING
        start = 7
        end = 7
        size = 0
```

Thus, to differentiate between `{ "", ""}` and the slightly-more-correct `{"": 1, "": 2}`, you need to check that you are getting a sequence of key-value pairs, not just strings.

In fact libjsmn is insane enough that if you parse `{ "key": 1 2 3 }` you get a key with multiple values.

```
[0]
        type = JSMN_OBJECT
        start = 0
        end = 16
        size = 1
[1]
        type = JSMN_STRING
        start = 3
        end = 6
        size = 3  # <--------------- notice this
[2]
        type = JSMN_PRIMITIVE
        start = 9
        end = 10
        size = 0
[3]
        type = JSMN_PRIMITIVE
        start = 11
        end = 12
        size = 0
[4]
        type = JSMN_PRIMITIVE
        start = 13
        end = 14
        size = 0
```